src/history/hs.c

```diff
@@ -1765,38 +1757,46 @@ __verify_history_store_id(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, uint32
         if (btree_id != this_btree_id)
             break;
 
+        /*
+         * If the stop time pair on the tombstone in the history store is already globally visible
+         * we can skip it.
+         */
+        if (__wt_txn_tw_stop_visible_all(session, &hs_cbt->upd_value->tw)) {
+            WT_STAT_CONN_INCR(session, cursor_next_hs_tombstone);
+            continue;
+        }
+
         /*
          * If we have already checked against this key, keep going to the next key. We only need to
          * check the key once.
          */
-        WT_ERR(__wt_compare(session, NULL, &hs_key, prev_hs_key, &cmp));
+        WT_ERR(__wt_compare(session, NULL, &key, prev_key, &cmp));
         if (cmp == 0)
             continue;
-        WT_WITH_PAGE_INDEX(session, ret = __hs_row_search(cbt, &hs_key, false));
-        WT_ERR(ret);
 
-        found = cbt->compare == 0;
-        WT_ERR(__cursor_reset(cbt));
+        WT_WITH_PAGE_INDEX(session, ret = __wt_row_search(cbt, &key, false, NULL, false, NULL));
```

Comment:
It's unusual to use a btree cursor directly, which I think leads to making it easier to get this wrong. I went looking at where the btree cursor comes from, and I think there is an issue there. It's directly calling __wt_btcur_open instead of going through the standard cursor create call. That means the code is bypassing protections we have around ensuring that dhandles are available and don't get closed underneath a cursor.

Can you look at changing that to create a full cursor instead of a btree cursor directly, and passing that into this function with a useful parameter name instead?